node/pkg/driver/device_connectivity/device_connectivity_iscsi_test.go

```diff
@@ -0,0 +1,203 @@
+						stdOut:   []byte("session command failure details"),
+						err:      errors.New("session command failed"),
+						exitCode: -1,
+					},
+				},
+			},
+		},
+		{
+			name:            "Should not login when session output is invalid",
+			portalsByTarget: portalsByTarget,
+			executeWithTimeoutCalls: []ExecuteWithTimeoutCall{
+				{
+					input: executeIscsiSessionArguments,
+					output: ExecuteWithTimeoutReturn{
+						stdOut: []byte("invalid session output"),
+					},
+				},
+			},
+		},
+		{
+			name:            "Should not login when all relevant portals are already in session output",
+			portalsByTarget: portalsByTarget,
+			executeWithTimeoutCalls: []ExecuteWithTimeoutCall{
+				{
+					input: executeIscsiSessionArguments,
+					output: ExecuteWithTimeoutReturn{
+						stdOut: []byte(twoPortalsIscsiSessionOutput),
+					},
+				},
+			},
+		},
+		{
+			name:            "Should not login when the only discover call fails",
+			portalsByTarget: portalsByTarget,
+			executeWithTimeoutCalls: []ExecuteWithTimeoutCall{
+				{
+					input: executeIscsiSessionArguments,
+					output: ExecuteWithTimeoutReturn{
+						stdOut: []byte(singlePortalIscsiSessionOutput),
+					},
+				},
+				{
+					input: executeIscsiDiscoverPortal2Arguments,
+					output: ExecuteWithTimeoutReturn{
+						stdOut: []byte("error"),
+						err:    errors.New("discover failed"),
+					},
+				},
+			},
+		},
+		{
+			name:            "Should login all portals not in session when a target discover call succeeds",
+			portalsByTarget: portalsByTarget,
+			executeWithTimeoutCalls: []ExecuteWithTimeoutCall{
+				{
+					input: executeIscsiSessionArguments,
+					output: ExecuteWithTimeoutReturn{
+						stdOut:   []byte{},
+						err:      errors.New("no sessions"),
+						exitCode: iSCSIErrNoObjsFound,
+					},
+				},
+				{
+					input: executeIscsiDiscoverPortal1Arguments,
+					output: ExecuteWithTimeoutReturn{
+						stdOut: []byte("error"),
+						err:    errors.New("discover failed"),
+					},
+				},
+				{
+					input: executeIscsiDiscoverPortal2Arguments,
+				},
+				{
+					input: executeIscsiLoginPortal1Arguments,
+				},
+				{
+					input: executeIscsiLoginPortal2Arguments,
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+
+		t.Run(tc.name, func(t *testing.T) {
+			mockCtrl := gomock.NewController(t)
+			defer mockCtrl.Finish()
+
+			fakeExecuter := mocks.NewMockExecuterInterface(mockCtrl)
+			fakeHelper := mocks.NewMockOsDeviceConnectivityHelperScsiGenericInterface(mockCtrl)
+
+			for _, c := range tc.executeWithTimeoutCalls {
+				fakeExecuter.EXPECT().ExecuteWithTimeout(c.input.mSeconds, c.input.command, c.input.commandArguments).Return(
+					c.output.stdOut, c.output.err)
+				if c.output.err != nil && c.output.exitCode != 0 {
+					fakeExecuter.EXPECT().GetExitCode(gomock.Any()).Return(c.output.exitCode, true)
+				}
+			}
+
+			o := NewOsDeviceConnectivityIscsiForTest(fakeExecuter, fakeHelper)
+			o.EnsureLogin(tc.portalsByTarget)
```

Comment:
fail login not blocking us?

Reply:
nope

Reply:
why?

Reply:
we never added a logic that can tell whether the final connection state is "good enough" (I don't think it's trivial).
currently if the final connection state is not "good enough", it will cause failure later in the flow, same as before the login feature

Reply:
you could call it a "best effort, no guarantee" feature

Reply:
ok